pkl-gradle/src/main/java/org/pkl/gradle/task/EvalTask.java

```diff
@@ -39,27 +48,44 @@ public abstract class EvalTask extends ModulesTask {
   @Optional
   public abstract Property<String> getModuleOutputSeparator();
 
-  @OutputDirectory
-  @Optional
-  public abstract DirectoryProperty getMultipleFileOutputDir();
-
   @Input
   @Optional
   public abstract Property<String> getExpression();
 
+  @Inject
+  public abstract ObjectFactory getObjectFactory();
+
+  private CliEvaluator getCliEvaluator() {
+    return new CliEvaluator(
+        new CliEvaluatorOptions(
+            getCliBaseOptions(),
+            getOutputFile().get().getAsFile().getAbsolutePath(),
+            getOutputFormat().get(),
+            getModuleOutputSeparator().get(),
+            mapAndGetOrNull(getMultipleFileOutputDir(), (it) -> it.getAsFile().getAbsolutePath()),
+            getExpression().get()));
+  }
+
+  @OutputFiles
+  @Optional
+  public SetProperty<File> getOutputPaths() {
+    var ret = getObjectFactory().setProperty(File.class);
+    ret.value(getCliEvaluator().getOutputFiles()).disallowChanges();
+    return ret;
+  }
```

Comment:
> Unfortunately, Gradle's model of input and output properties and their caching expects that all properties have statically known values after configuration time

This can't be true. Surely a task can take a `FileTree` as input whose files don't exist at configuration time?

> Pkl evaluation can depend on external, non-file inputs, like environment variables. 

I don't think that's an issue here because all such inputs are configured through the task (see `BasePklTask`, etc.).

Evaluating Pkl files at configuration time is not OK. But looking at `CliEvaluator`, I think that "computing output paths from a pattern does not, in fact, require Pkl evaluation" is correct.

Reply:
> Surely a task can take a FileTree as input whose files don't exist at configuration time?

Yes, it most definitely can, but it does not contradict what I'm saying. I might not have expressed it fully clearly, but essentially what I wanted to say is that in Gradle model, file collections must be fully configured before the task execution time. It does not mean that their contents can't be dynamic - they most certainly can, like class files produced by compilation tasks, and they can be absent before some task is executed - but that their configuration can't be dynamic, i.e. you cannot add new sources to a file collection or a file tree at task execution time.

When I was implementing the first version of the evaluation task, I assumed that it is not possible to determine exact file names and directory names from patterns without evaluation, i.e. without executing the task logic first. Because patterns are flexible, it would've been impossible to statically configure a `FileCollection` representing output files/directories, and therefore I had to make the tasks untracked.

Since apparently it _is_ possible to resolve patterns and determine outputs without running evaluation first, therefore, it should be possible to determine the full set of output files/directories at configuration time, and therefore we can make the task tracked.

> I don't think that's an issue here because all such inputs are configured through the task (see superclasses of EvalTask).

Fair point, forgot that these items were there. Yes, in this case evaluation should be self-contained and thus cached.